File: scripts/docs/validate-lifecycle-docs.py
#!/usr/bin/env python3
from __future__ import annotations

import re
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
DOCS_ROOT = ROOT / "docs"

KIND_CONFIG = {
    "planning": {
        "base": DOCS_ROOT / "planning",
        "statuses": {"proposed", "approved", "archived"},
        "required": {"id", "title", "status", "doc_type", "prd_mode", "target_branch", "owner"},
        "doc_types": {"feature", "architecture", "workflow", "runbook", "release-note", "adr", "other"},
        "prd_modes": {"none", "generate", "update"},
        "link_field": "linked_prd",
    },
    "prds": {
        "base": DOCS_ROOT / "prds",
        "statuses": {"draft", "approved", "archived"},
        "required": {"id", "title", "status", "doc_type", "source_doc", "target_branch", "owner"},
        "doc_types": {"prd"},
        "prd_modes": None,
        "link_field": "source_doc",
    },
}

ID_PATTERN = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


def parse_frontmatter(path: Path) -> tuple[dict[str, object], list[str]]:
    text = path.read_text(encoding="utf-8")
    lines = text.splitlines()
    if not lines or lines[0].strip() != "---":
        return {}, [f"{path}: missing frontmatter block"]

    data: dict[str, object] = {}
    idx = 1
    while idx < len(lines):
        line = lines[idx]
        if line.strip() == "---":
            return data, []
        if not line.strip():
            idx += 1
            continue
        if line.startswith("  - ") or line.startswith("\t- "):
            return {}, [f"{path}: invalid top-level list item in frontmatter"]
        if ":" not in line:
            return {}, [f"{path}: invalid frontmatter line '{line}'"]
        key, raw_value = line.split(":", 1)
        key = key.strip()
        value = raw_value.strip()
        if not key:
            return {}, [f"{path}: invalid empty frontmatter key"]
        if not value:
            items: list[str] = []
            idx += 1
            while idx < len(lines):
                nested = lines[idx]
                if nested.strip() == "---":
                    data[key] = items
                    return data, []
                if not nested.strip():
                    idx += 1
                    continue
                if nested.startswith("  - "):
                    items.append(nested[4:].strip())
                    idx += 1
                    continue
                break
            data[key] = items
            continue
        if value.startswith("[") and value.endswith("]"):
            inner = value[1:-1].strip()
            data[key] = [] if not inner else [item.strip().strip("'\"") for item in inner.split(",")]
        elif value.lower() in {"true", "false"}:
            data[key] = value.lower() == "true"
        else:
            data[key] = value.strip("'\"")
        idx += 1

    return {}, [f"{path}: missing closing frontmatter delimiter"]


def validate_file(kind: str, path: Path) -> list[str]:
    cfg = KIND_CONFIG[kind]
    frontmatter, parse_errors = parse_frontmatter(path)
    if parse_errors:
        return parse_errors

    errors: list[str] = []
    rel = path.relative_to(cfg["base"])
    parts = rel.parts
    if len(parts) < 2:
        errors.append(f"{path}: expected file inside status subdirectory")
        return errors

    status_dir = parts[0]
    if status_dir not in cfg["statuses"]:
        errors.append(f"{path}: unknown lifecycle status directory '{status_dir}'")
        return errors

    missing = sorted(field for field in cfg["required"] if not frontmatter.get(field))
    for field in missing:
        errors.append(f"{path}: missing required frontmatter field '{field}'")

    status = str(frontmatter.get("status", ""))
    if status and status != status_dir:
        errors.append(f"{path}: frontmatter status '{status}' does not match directory '{status_dir}'")

    doc_id = str(frontmatter.get("id", ""))
    if doc_id and not ID_PATTERN.match(doc_id):
        errors.append(f"{path}: id must be lowercase kebab-case")

    title = str(frontmatter.get("title", ""))
    if title and len(title) < 8:
        errors.append(f"{path}: title should be descriptive (minimum 8 characters)")

    doc_type = str(frontmatter.get("doc_type", ""))
    if doc_type and doc_type not in cfg["doc_types"]:
        errors.append(f"{path}: unsupported doc_type '{doc_type}'")

    owner = str(frontmatter.get("owner", ""))
    if owner and not owner.strip():
        errors.append(f"{path}: owner must be non-empty")

    target_branch = str(frontmatter.get("target_branch", ""))
    if target_branch and "/" in target_branch and target_branch.startswith("refs/"):
        errors.append(f"{path}: target_branch should be a branch name, not a full ref")

    if kind == "planning":
        prd_mode = str(frontmatter.get("prd_mode", ""))
        if prd_mode and prd_mode not in cfg["prd_modes"]:
            errors.append(f"{path}: unsupported prd_mode '{prd_mode}'")
        if status == "approved" and doc_type in {"feature", "architecture", "workflow"} and prd_mode == "none":
            errors.append(
                f"{path}: approved {doc_type} docs must declare prd_mode 'generate' or 'update'"
            )
        if status != "approved" and prd_mode == "update":
            errors.append(f"{path}: prd_mode 'update' is only valid for approved planning docs")

    if kind == "prds":
        if str(frontmatter.get("doc_type", "")) != "prd":
            errors.append(f"{path}: PRD documents must use doc_type 'prd'")
        source_doc = str(frontmatter.get("source_doc", ""))
        if source_doc and not source_doc.startswith("docs/planning/"):
            errors.append(f"{path}: source_doc must point to docs/planning/")

    return errors


def find_lifecycle_files() -> list[tuple[str, Path]]:
    files: list[tuple[str, Path]] = []
    for kind, cfg in KIND_CONFIG.items():
        if not cfg["base"].exists():
            continue
        for path in sorted(cfg["base"].rglob("*.md")):
            files.append((kind, path))
    return files


def main() -> int:
    errors: list[str] = []
    files = find_lifecycle_files()
    for kind, path in files:
        errors.extend(validate_file(kind, path))

    if errors:
        print("Lifecycle documentation validation failed:")
        for error in errors:
            print(f"  - {error}")
        return 1

    print(f"Lifecycle documentation validation passed for {len(files)} file(s).")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/docs/validate-lifecycle-docs.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
const docsRoot = path.join(root, "docs");
const markdownPattern = /\[[^\]]+\]\(([^)]+)\)/g;

const kindConfig = {
  planning: {
    base: path.join(docsRoot, "planning"),
    statuses: new Set(["proposed", "approved", "archived"]),
    required: new Set(["id", "title", "status", "doc_type", "prd_mode", "target_branch", "owner"]),
    docTypes: new Set(["feature", "architecture", "workflow", "runbook", "release-note", "adr", "other"]),
    prdModes: new Set(["none", "generate", "update"]),
  },
  prds: {
    base: path.join(docsRoot, "prds"),
    statuses: new Set(["draft", "approved", "archived"]),
    required: new Set(["id", "title", "status", "doc_type", "source_doc", "target_branch", "owner"]),
    docTypes: new Set(["prd"]),
    prdModes: null,
  },
};

const idPattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

function parseFrontmatter(filePath) {
  const text = fs.readFileSync(filePath, "utf8");
  const lines = text.split(/\r?\n/);
  if (lines.length === 0 || lines[0].trim() !== "---") {
    return [{}, [`${filePath}: missing frontmatter block`]];
  }

  const data = {};
  let index = 1;
  while (index < lines.length) {
    const line = lines[index];
    if (line.trim() === "---") {
      return [data, []];
    }
    if (!line.trim()) {
      index += 1;
      continue;
    }
    if (line.startsWith("  - ") || line.startsWith("\t- ")) {
      return [{}, [`${filePath}: invalid top-level list item in frontmatter`]];
    }
    if (!line.includes(":")) {
      return [{}, [`${filePath}: invalid frontmatter line '${line}'`]];
    }

    const [rawKey, ...rest] = line.split(":");
    const key = rawKey.trim();
    const value = rest.join(":").trim();
    if (!key) {
      return [{}, [`${filePath}: invalid empty frontmatter key`]];
    }

    if (!value) {
      const items = [];
      index += 1;
      while (index < lines.length) {
        const nested = lines[index];
        if (nested.trim() === "---") {
          data[key] = items;
          return [data, []];
        }
        if (!nested.trim()) {
          index += 1;
          continue;
        }
        if (nested.startsWith("  - ")) {
          items.push(nested.slice(4).trim());
          index += 1;
          continue;
        }
        break;
      }
      data[key] = items;
      continue;
    }

    if (value.startsWith("[") && value.endsWith("]")) {
      const inner = value.slice(1, -1).trim();
      data[key] = inner ? inner.split(",").map((item) => item.trim().replace(/^['"]|['"]$/g, "")) : [];
    } else if (value === "true" || value === "false") {
      data[key] = value === "true";
    } else {
      data[key] = value.replace(/^['"]|['"]$/g, "");
    }
    index += 1;
  }

  return [{}, [`${filePath}: missing closing frontmatter delimiter`]];
}

function listMarkdownFiles(baseDir) {
  if (!fs.existsSync(baseDir)) {
    return [];
  }
  const files = [];
  const stack = [baseDir];
  while (stack.length > 0) {
    const current = stack.pop();
    for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
      const fullPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        stack.push(fullPath);
        continue;
      }
      if (entry.isFile() && entry.name.endsWith(".md")) {
        files.push(fullPath);
      }
    }
  }
  files.sort();
  return files;
}

function validateFile(kind, filePath) {
  const cfg = kindConfig[kind];
  const [frontmatter, parseErrors] = parseFrontmatter(filePath);
  if (parseErrors.length > 0) {
    return parseErrors;
  }

  const errors = [];
  const rel = path.relative(cfg.base, filePath);
  const parts = rel.split(path.sep);
  if (parts.length < 2) {
    errors.push(`${filePath}: expected file inside status subdirectory`);
    return errors;
  }

  const statusDir = parts[0];
  if (!cfg.statuses.has(statusDir)) {
    errors.push(`${filePath}: unknown lifecycle status directory '${statusDir}'`);
    return errors;
  }

  for (const field of cfg.required) {
    if (!frontmatter[field]) {
      errors.push(`${filePath}: missing required frontmatter field '${field}'`);
    }
  }

  const status = String(frontmatter.status ?? "");
  if (status && status !== statusDir) {
    errors.push(`${filePath}: frontmatter status '${status}' does not match directory '${statusDir}'`);
  }

  const docId = String(frontmatter.id ?? "");
  if (docId && !idPattern.test(docId)) {
    errors.push(`${filePath}: id must be lowercase kebab-case`);
  }

  const title = String(frontmatter.title ?? "");
  if (title && title.length < 8) {
    errors.push(`${filePath}: title should be descriptive (minimum 8 characters)`);
  }

  const docType = String(frontmatter.doc_type ?? "");
  if (docType && !cfg.docTypes.has(docType)) {
    errors.push(`${filePath}: unsupported doc_type '${docType}'`);
  }

  const owner = String(frontmatter.owner ?? "");
  if (Object.hasOwn(frontmatter, "owner") && !owner.trim()) {
    errors.push(`${filePath}: owner must be non-empty`);
  }

  const targetBranch = String(frontmatter.target_branch ?? "");
  if (targetBranch && targetBranch.startsWith("refs/") && targetBranch.includes("/")) {
    errors.push(`${filePath}: target_branch should be a branch name, not a full ref`);
  }

  if (kind === "planning") {
    const prdMode = String(frontmatter.prd_mode ?? "");
    if (prdMode && !cfg.prdModes.has(prdMode)) {
      errors.push(`${filePath}: unsupported prd_mode '${prdMode}'`);
    }
    if (status === "approved" && ["feature", "architecture", "workflow"].includes(docType) && prdMode === "none") {
      errors.push(`${filePath}: approved ${docType} docs must declare prd_mode 'generate' or 'update'`);
    }
    if (status !== "approved" && prdMode === "update") {
      errors.push(`${filePath}: prd_mode 'update' is only valid for approved planning docs`);
    }
  }

  if (kind === "prds") {
    if (docType && docType !== "prd") {
      errors.push(`${filePath}: PRD documents must use doc_type 'prd'`);
    }
    const sourceDoc = String(frontmatter.source_doc ?? "");
    if (sourceDoc && !sourceDoc.startsWith("docs/planning/")) {
      errors.push(`${filePath}: source_doc must point to docs/planning/`);
    }
  }

  return errors;
}

function validateLocalLinks() {
  const errors = [];
  for (const filePath of listMarkdownFiles(docsRoot)) {
    const text = fs.readFileSync(filePath, "utf8");
    for (const match of text.matchAll(markdownPattern)) {
      const value = match[1].trim();
      if (!value || value.includes("://") || value.startsWith("#") || value.startsWith("mailto:")) {
        continue;
      }
      const localPath = value.split("#", 1)[0];
      const resolved = path.resolve(path.dirname(filePath), localPath);
      if (!fs.existsSync(resolved)) {
        errors.push(`${filePath}: broken local link '${value}'`);
      }
    }
  }
  return errors;
}

const files = [];
for (const [kind, cfg] of Object.entries(kindConfig)) {
  for (const filePath of listMarkdownFiles(cfg.base)) {
    files.push([kind, filePath]);
  }
}

const errors = [];
for (const [kind, filePath] of files) {
  errors.push(...validateFile(kind, filePath));
}
errors.push(...validateLocalLinks());

if (errors.length > 0) {
  console.log("Lifecycle documentation validation failed:");
  for (const error of errors) {
    console.log(`  - ${error}`);
  }
  process.exit(1);
}

console.log(`Lifecycle documentation validation passed for ${files.length} file(s).`);
